src/lib.rs

```diff
@@ -1,70 +1,85 @@
-                bail!("Got extra arguments, crate downloads take no extra arguments.");
-            }
-            clone_crate(name, version_req, &pkg_info)?;
+#[macro_use]
+extern crate log;
+
+static APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));
+
+/// https://api.bitbucket.org/2.0/repositories
+pub const DEFAULT_BITBUCKET_URL: &'static str = "https://api.bitbucket.org/2.0/repositories";
+/// https://github.com
+pub const DEFAULT_GITHUB_URL: &'static str = "https://github.com";
+/// https://gitlab.com
+pub const DEFAULT_GITLAB_URL: &'static str = "https://gitlab.com";
+/// https://crates.io
+pub const DEFAULT_REGISTRY_URL: &'static str = "https://crates.io";
+
+/// An enum representation of supported cloning methods.
+#[derive(Debug, Clone)]
+pub enum CloneMethodKind {
+    Crate,
+    Git,
+    Mercurial,
+    Pijul,
+    Fossil,
+    Auto,
+}
+
+impl CloneMethodKind {
+    /// Returns the underlying command line command for the method.
+    pub fn command(&self) -> &str {
+        match *self {
+            CloneMethodKind::Crate => "crate",
+            CloneMethodKind::Git => "git",
+            CloneMethodKind::Mercurial => "hg",
+            CloneMethodKind::Pijul => "pijul",
+            CloneMethodKind::Fossil => "fossil",
+            CloneMethodKind::Auto => "auto",
         }
-        "git" | "hg" | "pijul" | "fossil" => {
-            if let Some(version_req) = version_req {
-                bail!(
-                    "Specifying a version `{}` only works with the `crate` method.",
-                    version_req
-                );
-            }
-            run_clone(method, &repo, extra)?;
+    }
+
+    /// Creates a `CloneMethodKind` from a method name. If no name matches then None is returned.
+    /// Current options are `crate`, `git`, `hg`, `mercurial`, `pijul`, `fossil`, and `auto`
+    pub fn from(method_name: &str) -> Option<CloneMethodKind> {
+        match method_name {
+            "crate" => Some(CloneMethodKind::Crate),
+            "git" => Some(CloneMethodKind::Git),
+            "hg" => Some(CloneMethodKind::Mercurial),
+            "mercurial" => Some(CloneMethodKind::Mercurial),
+            "pijul" => Some(CloneMethodKind::Pijul),
+            "fossil" => Some(CloneMethodKind::Fossil),
+            "auto" => Some(CloneMethodKind::Auto),
+            _ => None,
         }
-        _ => bail!("Unsupported method `{}`", method),
     }
+}
+
+/// A struct containg all url and workspace information necessary to clone a crate.
+#[derive(Debug, Clone)]
+pub struct Cloner {
+    /// Defaults to https://crates.io
+    registry_url: String,
+
+    /// Defaults to https://github.com
+    github_url: String,
 
-    Ok(())
+    /// Defaults to https://gitlab.com
+    gitlab_url: String,
+
+    /// Defaults to https://api.bitbucket.org/2.0/repositories
+    bitbutcket_url: String,
+
+    /// Defaults to `std::env::current_dir()`
+    work_dir: PathBuf,
```

Comment:
Can this comment explain what the value is for?  Any maybe a name like `out_dir` would be more descriptive?

Reply:
Added a small comment and changed the name.